---
title: "ProMetIS: proteomics and metabolomics data integration"
author: "Alyssa Imbert and Etienne Thevenot (ProMetIS consortium)"
date: "`r doc_date()`"

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: "dubii_session4_ProMetIS_practical.bib"
output:
  BiocStyle::html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: false
---

![](figures/prometis_logo.png)

# Introduction

Genes are pleiotropic and getting a better knowledge of their function requires a comprehensive characterization of their mutants. Within the *ProMetIS* project, a multi-level dataset has been generated, which combines phenomic, proteomic and metabolomic acquisitions from liver tissues of a mouse knock-out model lacking the *Lat* (linker for activation of T cells) gene [@Imbert_PrometisDeepPhenotyping_]. The data and processing code are publicly available in the [ProMetIS](https://github.com/IFB-ElixirFr/ProMetIS) R package to ensure accessibility, interoperability, and reusability.

Here, we propose to analyse this dataset and characterize the impact of the deletion of the *Lat* gene.

# Loading the data

The multi-omics dataset is loaded as a MultiDataSet object.

```{r data_dir}
lat_mset_dir.c <- ProMetIS::aggregated_dir.c("LAT")
lat.mset <- phenomis::reading(lat_mset_dir.c, report.c = "none")
```

Let us view the content of this dataset:

```{r}
lat.mset
```

> _**Questions:**_
> 
> * How many datasets are included?
> 
> * How many technologies?
> 
> * How many tissues?
> 
> * How many samples?
> 
> * How many features?

# Restricting to the liver (annotated) datasets

We now restrict the dataset to the liver samples:

```{r liver}
# subsetting the MultiDataSet to the datasets containing 'liver' in their name
latliv.mset <- lat.mset[, grep("liver", names(lat.mset), value = TRUE)]
```

and to the annotated metabolomics features only:

```{r}
# for each dataset
for (set.c in grep("metabolomics", names(latliv.mset), value = TRUE)) {

  # extracting the ExpressionSet
  eset <- latliv.mset[[set.c]]

  # restricting to the features with a name
  eset <- eset[Biobase::fData(eset)[, "name"] != "", ]

  # updating the MultiDataSet
  stopifnot(methods::validObject(eset))

  latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)

}
```

Renaming the datasets

```{r}
latliv_oldname.mset <- latliv.mset
latliv.mset <- MultiDataSet::createMultiDataSet()

for (set.c in names(latliv_oldname.mset)) {

  # extracting the ExpressionSet
  eset <- latliv_oldname.mset[[set.c]]

  latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                        eset,
                                        dataset.type = switch(set.c,
                                                              proteomics_liver = "proteo",
                                                              metabolomics_liver_c18hypersil_pos = "metabo_c18",
                                                              metabolomics_liver_hilic_neg = "metabo_hilic"),
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)

}
latliv.mset <- latliv.mset[, c("proteo", "metabo_c18", "metabo_hilic")]
```


```{r}
latliv.mset
```

> _**Questions:**_
> 
> * How many datasets are included?
> 
> * How many technologies?
> 
> * How many tissues?
> 
> * How many samples?
> 
> * How many features?

# Data exploration (`phenomis::inspecting`)

The phenomis::inspecting method applied to a MultiDataSet (or ExpressionSet) provides
a numerical and graphical summary of the data:

```{r}
latliv.mset <- phenomis::inspecting(latliv.mset, report.c = "none")
```

> _**Questions:**_
> 
> * What shows each graphic?
> 
> * How many missing values are present in the datasets?
> 
> * How many zero values are present in the datasets?
> 
> * Have the intensities been log transformed?
> 
> * Are there outliers?


# Principal Component Analysis (`ropls::opls`)

We then focus on the Principal Component Analysis to see if we have clusters corresponding to the genotype.

Building the PCA model:

```{r pca_building}
latliv.pca <- ropls::opls(latliv.mset, info.txtC = "none")
```

> _**Questions:**_
> 
> * For each dataset, how many components contribute to more than 10% of the total variance?

Score-plot colored according to the genotype:

```{r pca_genotype}
# getting the genotype info in the sampleMetadata from one of the dataset
genotype.fc <- factor(Biobase::pData(latliv.mset[[names(latliv.mset)[1]]])[, "gene"])
ropls::plot(latliv.pca, typeVc = "x-score",
            parAsColFcVn = genotype.fc)
```

> _**Questions:**_
> 
> * Are the samples from the two genotypes clustered on the score-plot?
>
> * Which omic approach gives the best separation?
>
> * What happens with the protein data?

# Univariate Hypothesis Testing (`phenomis::hypotesting`)


```{r hypotesting}
latliv.mset <- phenomis::hypotesting(latliv.mset,
                                     factor_names.vc = c("gene", "sex"),
                                     test.c = "limma2waysInter",
                                     report.c = "none")
```

> _**Questions:**_
> 
> * Are there features with significant difference between means according to the genotype in all datasets?
>
> * Are there features with significant difference between means according to the sex in all datasets?
>
> * Are there interactions between genotype and sex?

# Multi-omics integration

## Multiple co-intertia analysis (**omicade4** package)

The following analysis is adapted from the [**omicade4** vignette](https://bioconductor.org/packages/release/bioc/vignettes/omicade4/inst/doc/omicade4.pdf) and the **Using MultiDataSet with third party R packages** additional file 2 [[@HernandezFerrer_MultidatasetRPackage_2017]](https://doi.org/10.1186/s12859-016-1455-1).

Restricting the datasets to common samples:

```{r mcia_common}
latliv.mset <- MultiDataSet::commonSamples(latliv.mset)
```

Exporting the data matrices:

```{r mcia_data}
latliv_mn.ls <- MultiDataSet::as.list(latliv.mset)
```

Replacing the "." by "_" in the metabolomics feature names (otherwise the names will be truncated during the call to mcia):

```{r mcia_feature_names}
latliv_mn.ls <- lapply(latliv_mn.ls, function(latliv.mn) {
  rownames(latliv.mn) <- gsub(".", "_", rownames(latliv.mn), fixed = TRUE)
  latliv.mn
})
```

Performing MCIA:

```{r mcia_compute}
latliv.mcia <- omicade4::mcia(latliv_mn.ls)
```

Plotting the results (samples are colored by genotype):

```{r mcia_plot}
gene.vc <- Biobase::pData(latliv.mset)[["proteomics_liver"]][, "gene"]
require(ade4)
omicade4::plot.mcia(latliv.mcia, axes = 1:2, phenovec = gene.vc, sample.lab = FALSE)
```

On the projection plot, axis 2 separates LAT and WT genotypes, whereas females and males are separated by axis 1.

Variables most contributing to axis 2 are metabolites:

```{r selectVar2}
omicade4::selectVar(latliv.mcia, a1.lim = c(-Inf, Inf), a2.lim = c(2, Inf))
```
Contribution to axis 1 also includes proteins:

```{r selectVar1}
omicade4::selectVar(latliv.mcia, a1.lim = c(2.5, Inf), a2.lim = c(-Inf, Inf))
```

## The mixOmics software

[@Rohart_MixomicsRPackage_2017](https://doi.org/10.1371/journal.pcbi.1005752)

```{r data_import, message=FALSE}
exprs.ls <- MultiDataSet::as.list(latliv.mset)
texprs.ls <- lapply(exprs.ls, t)

## Design matrix

# We create the design matrix. The weights were set to 0.1 between the blocks of omics and to 1 between the blocks of omics and the response block.

design.mn <- matrix(0.1,
                    ncol = length(exprs.ls),
                    nrow = length(exprs.ls), 
                    dimnames = list(names(exprs.ls),
                                    names(exprs.ls)))

diag(design.mn) <- 0

varsel.ls <- list(proteo = c(20,20),
                  metabo_c18 = c(10, 20),
                  metabo_hilic = c(60, 60))

# list.keepX_FW = list()

res.block.spls <- mixOmics::block.splsda(X = texprs.ls,
                                         Y = genotype.fc,
                                         design = design.mn,
                                         scheme = 'horst',
                                         keepX = varsel.ls)
```

### Scores

```{r plotDiablo}
mixOmics::plotDiablo(res.block.spls)
```

### Loadings

```{r}
mixOmics::plotLoadings(res.block.spls,
                       comp = 1,
                       legend.color = ProMetIS::palette.vc()[c("WT", "LAT")],
                       contrib = 'max', method = 'median')
```

### Plot individuals

```{r plotIndiv}
mixOmics::plotIndiv(res.block.spls,
                                           ind.names = TRUE, 
                                           legend = TRUE,
                                           ellipse = TRUE,
                                           col.per.group = ProMetIS::palette.vc()[c("WT", "LAT")])
```

```{r plotArrow}
plot_arrow <- lapply(names(sgccda.ls),
                     function(gene_tissue.c)
                       mixOmics::plotArrow(sgccda.ls[[gene_tissue.c]][["block_splsda"]],
                                           ind.names = FALSE, 
                                           col = ProMetIS::palette.vc()[as.character(sgccda.ls[[gene_tissue.c]][["gene.fc"]])],
                                           title = gene_tissue.c))
```

### Plot variables

```{r plotVar}
mixOmics::plotVar(res.block.spls,
                                       var.names = FALSE, 
                                       style = 'graphics', 
                                       legend = TRUE)
```

### Circos

```{r circosPlot}
mixOmics::circosPlot(res.block.spls,
                                             cutoff = 0.7,
                                             line = TRUE,
                                             comp = 1,
                                             color.Y = ProMetIS::palette.vc()[c("WT", "LAT")])

```

### Heatmap

```{r cimDiablo}
mixOmics::cimDiablo(res.block.spls,
                    comp = 1,
                    color.Y = ProMetIS::palette.vc()[c("WT", "LAT")])
```


### Network

```{r}
mixOmics::network(res.block.spls,
                                       color.node = c("#1F78B4", '#33A02C'),
                                       shape.node = c("rectangle", "rectangle"),
                                       lty.edge = "solid", lwd.edge =  1,
                                       cutoff = 0.75)
```

### Integrating the variable selection results in the dataset

```{r}

for (set.c in names(latliv.mset)) {
  
  eset <- latliv.mset[[set.c]]
  
  fdata.df <- Biobase::fData(eset)
    
  loadings.ls <- res.block.spls[["loadings"]]
  
  loadings.vn <- loadings.ls[[set.c]][, "comp1"]
  
  loadings.vn <- loadings.vn[abs(loadings.vn) > 0]
  
  mixomics.vn <- numeric(nrow(fdata.df))
  names(mixomics.vn) <- rownames(fdata.df)
  
  stopifnot(all(names(loadings.vn) %in% names(mixomics.vn)))
  
  mixomics.vn[names(loadings.vn)] <- loadings.vn
  
  fdata.df[, "mixOmics"] <- mixomics.vn
  
  Biobase::fData(eset) <- fdata.df
  
  stopifnot(methods::validObject(eset))
  
  latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                        eset,
                                        dataset.type = set.c,
                                        GRanges = NA,
                                        overwrite = TRUE,
                                        warnings = FALSE)
  
}
```



# Appendix

## 3 tabular file format used for import/export

Each dataset consists of:

1. a numeric matrix of intensities (**dataMatrix**)

2. a data frame of sample metadata (**sampleMetadata**)

3. a data frame of variable metadata (**variableMetadata**)

Theses 3 tables can be conveniently imported to/exported from R as tabular files:

![](figures/phenomis_3table_format.png)

### **ExpressionSet** class for a single dataset

The following table describes useful Biobase methods for the handling of ExpressionSet objects (see the ['An introduction to Biobase and ExpressionSets'](https://bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf) documentation from the [**Biobase**](https://doi.org/doi:10.18129/B9.bioc.Biobase) package): 

| Biobase methods | Description |
|---|---|
| **dims(eset)** | 2-row numeric matrix of 'Features' and 'Samples' dimensions|
| **exprs(eset)** | 'variable times samples' numeric matrix - dataMatrix |
| **pData(eset)** | sample metadata data frame - sampleMetadata |
| **fData(eset)** | variable metadata data frame - variableMetadata |
| **sampleNames(eset)** | sample names |
| **featureNames(eset)** | variable names |
| **varLabels(eset)** | Column names of the sampleMetadata, pData(eset) |
| **fvarLabels(eset)** | Column names of the variableMetadata, fData(eset) |

Creating an ExpressionSet with `Biobase::ExpressionSet`:

```{r eset_create}
lat_dir.c <- ProMetIS::aggregated_dir.c("LAT")
latlivprot_dir.c <- file.path(lat_dir.c, "proteomics_liver")
dataMatrix.mn <- as.matrix(read.table(file.path(latlivprot_dir.c, "dataMatrix.tsv"),
                                      header = TRUE,
                                      row.names = 1,
                                      sep = "\t",
                                      comment.char = "",
                                      quote = "\""))
sampleMetadata.df <- read.table(file.path(latlivprot_dir.c, "sampleMetadata.tsv"),
                                header = TRUE,
                                row.names = 1,
                                sep = "\t",
                                comment.char = "",
                                quote = "\"")
variableMetadata.df <- read.table(file.path(latlivprot_dir.c, "variableMetadata.tsv"),
                                  header = TRUE,
                                  row.names = 1,
                                  sep = "\t",
                                  comment.char = "",
                                  quote = "\"")
latlivprot.eset <- Biobase::ExpressionSet(assayData = dataMatrix.mn,
                                          phenoData = Biobase::AnnotatedDataFrame(data = sampleMetadata.df),
                                          featureData = Biobase::AnnotatedDataFrame(data = variableMetadata.df),
                                          experimentData = Biobase::MIAME(title = "lat_proteomics_liver"))
```

which is equivalent to:

```{r eset_read}
latlivprot.eset <- phenomis::reading(latlivprot_dir.c)
```

Brackets may be used to restrict the ExpressionSet to specific sample or features:

```{r eset_subset}
wt.eset <- latlivprot.eset[, grep("W", Biobase::sampleNames(latlivprot.eset))]
Biobase::dims(wt.eset)
signif.eset <- latlivprot.eset[Biobase::fData(latlivprot.eset)[, "WT.KO_signif"] > 0, ]
Biobase::dims(signif.eset)
```

Exporting the ExpressionSet in the 3 table format:

```{r eset_write, eval = FALSE}
phenomis::writing(latlivprot.eset, dir.c = "your_prefered_directory")
```


### **MultiDataSet** class for multiple datasets

The following table describes useful MultiDataSet (and Biobase) methods for the handling of MultiDataSet objects [[@HernandezFerrer_MultidatasetRPackage_2017]](https://doi.org/10.1186/s12859-016-1455-1):

| MultiDataSet and Biobase methods | Description |
|---|---|
| **MultiDataSet::length(mset)** | number of datasets |
| **MultiDataSet::names(eset)** | names of the datasets |
| **Biobase::dims(mset)** | list of 2-element numeric named ('Features' and 'Samples') vectors of data set dimensions |
| **MultiDataSet::commonSamples(mset)** | multi-dataset restricted to the common samples |
| **MultiDataSet::as.list(mset)** | list of data matrices |
| **Biobase::pData(mset)** | list of sample metadata data frames |
| **Biobase::fData(mset)** | list of variable metadata data frames |
| **BiocGenerics::subset(mset)** | subsetting on common columns from the sampleMetadata and/or variableMetadata |

Creating a MultiDataSet with `MultiDataSet::createMultiDataSet`:

```{r mset_create}
lat_dir.c <- ProMetIS::aggregated_dir.c("LAT")
proteo.eset <- phenomis::reading(file.path(lat_dir.c, "proteomics_liver"), report.c = "none")
metabo_c18.eset <- phenomis::reading(file.path(lat_dir.c, "metabolomics_liver_c18hyper_pos"), report.c = "none")
metabo_hilic.eset <- phenomis::reading(file.path(lat_dir.c, "metabolomics_liver_hilic_neg"), report.c = "none")

latliv.mset <- MultiDataSet::createMultiDataSet()
latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                      proteo.eset,
                                      dataset.type = "proteomics_liver",
                                      GRanges = NA)
latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                      metabo_c18.eset,
                                      dataset.type = "metabolomics_liver_c18hyper_pos",
                                      GRanges = NA)
latliv.mset <- MultiDataSet::add_eset(latliv.mset,
                                      metabo_hilic.eset,
                                      dataset.type = "metabolomics_liver_hilic_neg",
                                      GRanges = NA)
```

which is equivalent to:

```{r mset_read}
latliv.mset <- phenomis::reading(lat_dir.c,
                                 subsets.vc = c("proteomics_liver",
                                                "metabolomics_liver_c18hyper_pos",
                                                "metabolomics_liver_hilic_neg"),
                                 report.c = "none")
```

Brackets may be used to restrict the MultiDataSet to specific samples or datasets:

```{r mset_subset}
wt.mset <- latliv.mset[grep("W", Biobase::sampleNames(latliv.mset)[["proteomics_liver"]], value = TRUE), ]
Biobase::dims(wt.mset)

metabo.mset <- latliv.mset[, c("metabolomics_liver_c18hyper_pos", "metabolomics_liver_hilic_neg")]
names(metabo.mset)
```

Note: In case of the restriction to a single dataset, the single bracket (as shown above) will outpout a MultiDataSet object. To get an ExpressionSet, either add the drop = TRUE argument or use the double brackets:

```{r mset_to_eset}
proteo.eset <- latliv.mset[["proteomics_liver"]]
```

Subsetting based on (common) columns from the sampleMetadata and/or variableMetadata may be performed with `BiocGenerics::subset`:

```{r mset_subset_features}
female.mset <- BiocGenerics::subset(latliv.mset, , sex == "F")
# or
signif.mset <- BiocGenerics::subset(latliv.mset, WT.KO_signif > 0, )
```

Exporting the MultiDataSet into the 3 table format (each dataset will be written as 3 tables in a specific subfolder):

```{r mset_write, eval = FALSE}
phenomis::writing(latliv.mset, dir.c = "your_prefered_directory")
```

# References

